doc/bbcbasic_to_text.1
'\" t
.TH BBCBASIC_TO_TEXT 1 "2020-07-17 18:13 GMT" Youngman \" -*- nroff -*-

.SH NAME
bbcbasic_to_text \- convert a saved BBC BASIC program to text

.SH SYNOPSIS
.B bbcbasic_to_text
[\--dialect D] [\-\-listo L]
.B bbcbasic_to_text
[\--help]
.SH DESCRIPTION
The
.B bbcbasic_to_text
program reads a saved BBC BASIC file (in any of several dialects)
and writes the text of the program, as if by the use of the
BASIC command
.BR LIST .

.SH OPTIONS
The following options are understood:
.IP "\-d \fID\fR, \-\-dialect \fID\fR"
Specify which BBC BASIC dialect to assume the input file uses.
The default is
.BR 6502 .
See
.B BBC BASIC DIALECTS
for a list.

.IP \-\-help
Show a message explaining how to use the program.

.IP "\-l \fIN\fR, --listo \fIN\fR"
Set the
.B LIST
option bits to the value
.IR N .
See
.B LIST OPTION BITS
for an explanation of how
.I N
is interpreted.  The default value for
.I N
is 7 (though this may change in the future).

.SH "BBC BASIC DIALECTS"
The following dialect keywords can be used for
.I D
in the
.B --dialect
.I D
option.

.TS
tab(|);
lb lb lb
l l
l l.
Dialect Keyword   | What is it?
6502| BBC Microcomputer.
32000| Same as 6502.  BBC Microcomputer 32016 Tube.
Z80| BBC BASIC Z80 (as in e.g. the Sinclair Z88).
8086| Same as Z80.  BBC BASIC for DOS.
ARM| Acorn RiscOS BASIC V.
Windows|T{
BBC BASIC for Windows
.br
BBC BASIC for SDL 2.0 (including Linux and Mac OS X)
T}
Mac|T{
Apple Classic MacOS.  For compatibility with the the MacOS X
implementation, use the Windows dialect instead.
T}
.TE

.SH "LIST OPTION BITS"

The
.B LIST
option bits are specified as a decimal number, just as in the BBC
BASIC command
.BR LISTO .
With bit 0 as the least-significant, they have the following
meanings:

.TS
tab(|);
lb lb
n l.
Bit | Meaning
0   | When set, insert a space after the line number
1   | When set, indent the body of \fBFOR\fR ... \fBNEXT\fR loops.
2   | When set, indent the body of \fBREPEAT\fR ... \fBUNTIL\fR loops.
.TE
Commonly used values are 0 (no spaces), 1 (just insert a space after
the line number) and 7 (insert spaces for all the situations in the
table above).

There appears to be some variation in
how the
.B NEXT
or
.B UNTIL
is handled in  the implementation of bits 1 and 2.
On some versions of BBC BASIC, the
.B NEXT
or
.B UNTIL
has the same indentation as the loop body.  On otherd
it has the same indentation as the matching
.B FOR
or
.BR REPEAT .
This program takes the second approach.
Using
.B \-\-listo 7
might give the following result:
.EX
   10 FOR I = 1 TO 10
   20   PRINT I
   30 NEXT
.EE
.P
The same program listed with
.B \-\-listo 0
looks like this:
.EX
   10FOR I = 1 TO 10
   20PRINT I
   30NEXT
.EE
.P
There is currently no way to get this program to produce a listing
like this:
.EX
   10 FOR I = 1 TO 10
   20   PRINT I
   30   NEXT
.EE
.P
It is possible for the on-disc program to include spaces too, so even
with
.B \-\-listo 0
a program might still look like this:
.EX
   10FOR I = 1 TO 10
   20PRINT I
   30NEXT
   40FOR J = 1 TO 10
   50  PRINT J
   60 NEXT
.EE
This can be confusing.

.SH "EXIT STATUS"
The exit status is zero when everything has gone well and non-zero
when there was a problem.  Problems might inclue that the input file
is not encoded as a BBC BASIC program, its actual dialect doesn't
match what was specified, or that the command-line inclued a usage
error.

.SH "CONFORMING TO"

BBC BASIC implementations differ in some details, and in a number of
cases,
.B bbcbasic_to_text
will render a BBC BASIC file as text where it
would not in fact be accepted by the specified dialect of BBC BASIC.
Also, the input file is not actually being parsed, so
.B bbcbasic_to_text
will render programs that are not really valid.
Examples include:

.IP \(bu
In the 6502 dialect, line numbers greater than 32767 are
accepted by this program even though the BBC Micro will not produce or
accept such programs.

.IP \(bu
In all dialects, sequences of non-tokenised characters are accepted
and passed though with the result that
.B bbcbasic_to_text
will accept and render invalid programs such as
.EX
 10 DANCE AS IF NOBODY IS WATCHING
.EE

.SH BUGS
Please report bugs to james@youngman.org.

.SH EXAMPLE

To format a program saved by BBC BASIC for Windows 2.0 or for BBC
BASIC for SDL 2.0 with formatting like
.BR "LISTO 7" :

.EX
bbcbasic_to_text --dialect=Windows prog.bbc
.EE

To format a program saved by BBC BASIC for (Classic) Macintosh with
formatting like
.BR "LISTO 0" :

.EX
bbcbasic_to_text --dialect=Mac --listo=0 prog.bbc
.EE


.SH "SEE ALSO"
.UR http://www.bbcbasic.co.uk/bbcbasic.html
BBC BASIC Website
.UE
.br
\fBdfs\fP(1), \fBenscript\fP(1)


basic/bbcbasic_to_text.c
//
//   Copyright 2020 James Youngman
//
//   Licensed under the Apache License, Version 2.0 (the "License");
//   you may not use this file except in compliance with the License.
//   You may obtain a copy of the License at
//
//       http://www.apache.org/licenses/LICENSE-2.0
//
//   Unless required by applicable law or agreed to in writing, software
//   distributed under the License is distributed on an "AS IS" BASIS,
//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//   See the License for the specific language governing permissions and
//   limitations under the License.
//
#include <assert.h>            // for assert
#include <unistd.h>	       // for optarg, optind
#include <getopt.h>            // for getopt_long, option
#include <stdbool.h>           // for bool, false, true
#include <stdio.h>             // for fprintf, perror, stderr, NULL, fflush
#include <stdlib.h>            // for strtol
#include <string.h>            // for strcmp

#include "decoder.h"           // for decode_file, destroy_decoder, new_decoder
#include "tokens.h"            // for set_dialect, print_dialects, etc...


static bool usage(FILE *f, const char *progname)
{
  return fprintf(f, "usage: %s [--listo=N] [--dialect=NAME] [input-file]...\n"
		 "Use the option --help to see the program's usage in more detail.\n",
		 progname) >= 0;
}

static bool help(FILE *f, const char *progname)
{
  return fprintf(f, "usage: %s [--listo=N] [--dialect=NAME] [input-file]...\n"
		 "If no input-file is listed, issue a usage message and exit.\n"
		 "If input-file is \"-\", read standard input.\n"
		 "Valid values for --listo are 0..7 inclusive.\n"
		 "You can list valid dialect names by specifying --dialect=help.\n"
		 "If the option --help is given, this usage message is printed and "
		 "nothing else is done.\n",
		 progname) >= 0;
}

static bool set_listo(const char *s, int *listo)
{
  long converted;
  char *end;

  /* There is no need to clear errno to distinguish LONG_MAX from
     overflow, since LONG_MAX is already greater than 7 on all ISO C
     compiant platforms.   Likewise 0 and LONG_MIN. */
  converted = strtol(s, &end, 10);
  if (s == end)
    {
      fprintf(stderr, "Value %s should be an integer.\n", s);
      return false;
    }
  if (*end)
    {
      fprintf(stderr, "Value %s should be an integer, but it was followed by trailing junk %s.\n", s, end);
      return false;
    }
  if (converted >= 0 && converted <= 7)
    {
      *listo = (int)converted;
      return true;
    }
  fprintf(stderr, "Value %s is out of range; the valid range is 0 to 7.\n", s);
  return false;
}


int wrapped_main(int argc, char *argv[])
{
  int exitval = 0;
  unsigned dialect;
  const char* default_dialect_name = "6502";
  const char *progname = argv[0];
  int listo = 7;
  int longindex;
  const struct option opts[] =
    {
     { "dialect", 1, NULL, 'd' },
     { "help", 0, NULL, 'h'},
     { "listo", 1, NULL, 'l' },
     { "dump-token-maps", 0, NULL, 'D' },
     { NULL, 0, NULL, 0 },
    };
  if (progname == NULL)
    {
      /* argv[0] can be NULL, you can achieve this with exec(). */
      progname = "bbcbasic_to_text";
    }
  assert(set_dialect(default_dialect_name, &dialect)); /* set the default */
  int opt;
  while ((opt=getopt_long(argc, argv, "+d:D:l:", opts, &longindex)) != -1)
    {
      switch (opt)
	{
	case '?':  // Unknown option or missing option argument.
	  // An error message was already issued.
	  usage(stderr, progname);
	  return 1;

	case 'D':
	  /* We don't document this option because we don't want to
	   * have to maintain backward compatibility in the output
	   * format.  This option is mainly for regression testing.
	   */
	  if (!internal_dump_all_dialects(optarg))
	    return 1;
	  return 0;

	case 'h':
	  if (!help(stdout, progname))
	    return 1;
	  return 0;

	case 'l':
	  if (!set_listo(optarg, &listo))
	    return 1;		/* error message already issued. */
	  break;

	case 'd':
	  if (0 == strcmp(optarg, "help"))
	    {
	      if (!print_dialects(stdout, default_dialect_name))
		return 1;
	    }
	  else if (!set_dialect(optarg, &dialect))
	    {
	      fprintf(stderr, "Unknown BASIC dialect '%s'\n", optarg);
	      print_dialects(stderr, default_dialect_name);
	      return 1;
	    }
	  break;
	}
    }

  if (optind == argc)
    {
      /* no input files listed. */
      fprintf(stderr, "You didn't specify any input files.\n");
      usage(stderr, progname);
      return 1;
    }
  for (; optind < argc; ++optind)
    {
      const char *name = argv[optind];
      FILE *f;
      if (0 == strcmp("-", argv[optind]))
	{
	  name = "standard input";
	  f = stdin;
	}
      else
	{
	  f = fopen(name, "rb");
	  if (NULL == f)
	    {
	      perror(name);
	      if (exitval < 1)
		exitval = 1;
	      continue;
	    }
	}
      struct decoder *dec = new_decoder(dialect, listo);
      if (0 == dec)
	{
	  fprintf(stderr, "failed to initialise decoder\n");
	  return 1;
	}

      if (!decode_file(dec, name, f))
	{
	  if (exitval < 1)
	    exitval = 1;
	}

      destroy_decoder(dec);
      dec = NULL;

      if (EOF == fclose(f))
	{
	  perror(name);
	  if (exitval < 1)
	    exitval = 1;
	}
    }
  return exitval;
}

int main(int argc, char *argv[])
{
  /* We use a wrapper like this to that various parts of the main
   * function can just "return 0" and still take advantage of the I/O
   * failure detection below.
   */
  int exitval = wrapped_main(argc, argv);

  /* Our formatted program went to stdout, so there is likely some
   * buffered data to be flushed.  The C library's exit handling would
   * flush it, but if those operations fail they won't change the exit
   * status of the program.  We, on the other hand, would want to let
   * the user know there was a problem, so we flush the data
   * ourselves.
   */
  if (0 != fflush(stdout))
    {
      perror("stdout");
      exitval = 1;
    }
  if (0 != fflush(stderr))
    {
      perror("stderr");  /* likely useless, but worthwhile attempt */
      exitval = 1;       /* but this is useful */
    }
  return exitval;
}
